File: mobile/android/feature/addressbook/src/main/kotlin/com/qervon/features/addressbook/AddressBookViewModel.kt
// =============================================================================
// File:           mobile/android/feature/addressbook/src/main/kotlin/com/qervon/features/addressbook/AddressBookViewModel.kt
// Project:        Qervon
// Author:         USDTG GROUP TECHNOLOGY LLC
// Developer:      Irfan Gedik
// Created Date:   2026-08-12
// Version:        0.1.0
//
// Description:
//   Loads/mutates the customer's saved addresses
//   (`POST/DELETE /v1/customer/profile/addresses`).
//
// License:
//   Qervon License v1.0 — see LICENSE in the repository root.
// =============================================================================

package com.qervon.features.addressbook

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.qervon.core.common.QervonApiException
import com.qervon.core.common.model.SavedAddress
import com.qervon.core.network.QervonApi
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class AddressBookUiState(
    val addresses: List<SavedAddress> = emptyList(),
    val isLoading: Boolean = false,
    val errorMessage: String? = null,
)

@HiltViewModel
class AddressBookViewModel @Inject constructor(private val api: QervonApi) : ViewModel() {

    private val _uiState = MutableStateFlow(AddressBookUiState())
    val uiState: StateFlow<AddressBookUiState> = _uiState.asStateFlow()

    fun refresh() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, errorMessage = null)
            try {
                val profile = api.getCustomerProfile()
                _uiState.value = _uiState.value.copy(addresses = profile.addresses)
            } catch (error: QervonApiException) {
                _uiState.value = _uiState.value.copy(errorMessage = error.message)
            } finally {
                _uiState.value = _uiState.value.copy(isLoading = false)
            }
        }
    }

    fun addAddress(label: String, latitude: Double, longitude: Double, fullAddress: String) {
        viewModelScope.launch {
            try {
                val profile = api.addAddress(label, latitude, longitude, fullAddress)
                _uiState.value = _uiState.value.copy(addresses = profile.addresses)
            } catch (error: QervonApiException) {
                _uiState.value = _uiState.value.copy(errorMessage = error.message)
            }
        }
    }

    fun removeAddress(addressId: String) {
        viewModelScope.launch {
            try {
                val profile = api.removeAddress(addressId)
                _uiState.value = _uiState.value.copy(addresses = profile.addresses)
            } catch (error: QervonApiException) {
                _uiState.value = _uiState.value.copy(errorMessage = error.message)
            }
        }
    }
}

File: mobile/ios/Features/QervonFeatures/Sources/AddressBookFeature/AddressBookViewModel.swift
// =============================================================================
// File:           mobile/ios/Features/QervonFeatures/Sources/AddressBookFeature/AddressBookViewModel.swift
// Project:        Qervon
// Author:         USDTG GROUP TECHNOLOGY LLC
// Developer:      Irfan Gedik
// Created Date:   2026-08-12
// Version:        0.1.0
//
// Description:
//   Loads and mutates the customer's saved address book
//   (`GET/POST/DELETE /v1/customer/profile/addresses`).
//
// License:
//   Qervon License v1.0 — see LICENSE in the repository root.
// =============================================================================

import Foundation
import QervonCore
import QervonNetworking

@MainActor
public final class AddressBookViewModel: ObservableObject {
    @Published public private(set) var addresses: [SavedAddress] = []
    @Published public private(set) var isLoading = false
    @Published public var errorMessage: String?

    private let api: QervonAPI

    public init(api: QervonAPI) {
        self.api = api
    }

    public func load() async {
        isLoading = true
        errorMessage = nil
        defer { isLoading = false }
        do {
            let profile = try await api.getCustomerProfile()
            addresses = profile.addresses
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    public func addAddress(label: String, coordinate: GeoLocation, fullAddress: String) async {
        errorMessage = nil
        do {
            let profile = try await api.addAddress(
                label: label,
                latitude: coordinate.latitude,
                longitude: coordinate.longitude,
                fullAddress: fullAddress
            )
            addresses = profile.addresses
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    public func removeAddress(_ address: SavedAddress) async {
        errorMessage = nil
        do {
            let profile = try await api.removeAddress(id: address.id)
            addresses = profile.addresses
        } catch {
            errorMessage = error.localizedDescription
        }
    }
}
